Include force:source:deploy errors in problem view

ChuckJonas opened the pull request:
### What does this PR do?
Shows compile errors in problem view (apex only!).  

### What issues does this PR fix or reference?
#588

ChuckJonas commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
     return commandBuilder.build();
   }
+
+  public execute(response: ContinueResponse<SelectedPath>): void {
```
this was a bit weird to me, overriding the execute statements means you pretty much loose all the "sandwich" code... would be nice if it instead returned the `execution` instead of `void` so you could extend without changing the default behavior...

```
let execution = super.execute(response);
execution.processExitSubject.subscribe(async exitCode => {
//....
});
```

ChuckJonas commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+  private createDiagnosticFromResult(err: Result) {
+    const ln = Number(err.lineNumber) - 1;
+    const col = Number(err.columnNumber) - 1;
+    const range = new vscode.Range(new vscode.Position(ln, col), new vscode.Position(ln, col));
```
salesforce only returns a position, not a range

ChuckJonas commented:
@ntotten  @vazexqi 
I think there is still more to do on this, but I would love a quick review.  I wasn't sure about some of the coding conventions... Tried to follow best what I saw in a few of the other places I looked.  

Also, I couldn't get the test to run in my env

ChuckJonas commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+import { SfdxCommandlet, SfdxCommandletExecutor, SfdxWorkspaceChecker } from './commands';
+import { FileType, ManifestOrSourcePathGatherer, SelectedPath } from './forceSourceRetrieve';
+
+interface Result {
```
Also, these types are only for apex and I'm not sure they are even complete in that regard...

lcampos replied:
I think these interfaces and the parsing methods should be moved to the `salesforcedx-utils-vscode` module under `cli` since we can potentially use them for parsing results from other command executions.

As mentioned below, I've tested errors on different metadata error results but will get back to you on what changes might be needed for the interfaces.

ChuckJonas commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private getDeployResultData(stdErr: string) {
```
will likely need to expand for different meta-data types

lcampos replied:
CLI output will always be in the format `{ status: 0|1, result: /* command execution result */ }`. I think just doing a JSON.parse (with some error handling) might be enough.

Something I'm still looking into is if all possible errors are posted in the shape of the `Result` & `ForceSourceDeployErrorResult` interfaces you've added. So far they work for apex, custom objects and aura (with some info missing that might be a bug/gap). I'll get back to you once I'm able to confirm this.

vazexqi replied:
Luis is right here. There are some deploy errors that do not conform to the shape. We have to do a best case approximation – what happens if there is no file URI (or no line/column) to associate with those.

ChuckJonas replied:
Awesome, ya looks like it does work with objects deployments as well.

As mentioned below, I'm struggling to find a simple & consistent way to only grab the response JSON from stderr.  Here's an example stderr from a object deploy command.  You can see it has extra junk before the response. 

```
▸    sfdx-cli: update available from 6.27.0 to 6.31.0-32aa4b4408
[xmldom error]	end tag name: type maybe not complete 
@#[line:9,col:15]
{"message":"Deploy failed.","status":1,"stack":"DeployFailed: Deploy failed.\n    at ALMError (/Users/jonas/.local/share/sfdx/plugins/node_modules/salesforce-alm/dist/lib/almError.js:44:19)\n    at MetadataRegistry.initializeMetadataTypeInfos.then.catch.e (/Users/jonas/.local/share/sfdx/plugins/node_modules/salesforce-alm/dist/lib/source/sourceApiCommand.js:58:31)\n    at tryCatcher (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/util.js:16:23)\n    at Promise._settlePromiseFromHandler (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:510:31)\n    at Promise._settlePromise (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:567:18)\n    at Promise._settlePromise0 (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:612:10)\n    at Promise._settlePromises (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:687:18)\n    at Async._drainQueue (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:138:16)\n    at Async._drainQueues (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:148:10)\n    at Immediate.Async.drainQueues [as _onImmediate] (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:17:14)\n    at runCallback (timers.js:773:18)\n    at tryOnImmediate (timers.js:734:5)\n    at processImmediate [as _immediateCallback] (timers.js:711:5)","name":"DeployFailed","result":[{"columnNumber":"28","lineNumber":"13","error":"Error parsing file: 'asdasd' is not a valid value for the enum 'ActionOverrideType' (13:28)","fullName":"abc__c","type":"CustomObject","filePath":"force-app/main/default/objects/abc__c/abc__c.object-meta.xml"}],"warnings":[]}
```

vazexqi replied:
We usually use `--loglevel fatal` to silence any other information (see https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-utils-vscode/src/cli/commandBuilder.ts#L63)

@amphro - Any thoughts on `[xmldom error]	end tag name: type maybe not complete `? I've not seen that one before.

ChuckJonas replied:
I'm using with `withJson()` which sets the `--loglevel fatal`, but I'm still getting the update and xml parsing junk... is this a bug with the CLI?

amphro replied:
I haven't seen `[xmldom error]	end tag name: type maybe not complete 
@#[line:9,col:15]` but the update available is well known and doesn't respect the `--loglevel fatal` either. We are picking up the story to always have the json response in stdout, which will simplify a lot of this. It should be fixed in the next month with an envvar to enable the new functionality to allow tools to slowly move over. 

Unfortunately, until that happens, there isn't a great workaround short of stipping everything before the first `{:` and hoping that is good enough.

lcampos commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
```
`workspace.rootPath` is deprecated (https://github.com/Microsoft/vscode/wiki/Extension-Authoring:-Adopting-Multi-Root-Workspace-APIs), you can use `workspace.workspaceFolders[0]` for now since we need to make further changes for the extensions to support multiple workspaces.

lcampos commented:
These changes also need tests, you can look at `forceSourceDeploy.test.ts` to see how we are currently writing tests.

Thanks for contributing to the project and feel free to ping me if you have questions/comments, looking forward to getting this to a point in which can be merged.

vazexqi commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+
+export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<SelectedPath> {
+
+  public static _errorCollection = vscode.languages.createDiagnosticCollection('deploy-errors');
```
Let's not use the `_` in front of variables names.

ChuckJonas replied:
whats your preferred naming convention for statics?

vazexqi commented on packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts:
```diff
@@ -5,37 +5,129 @@
+  }
+
+  private getDeployResultData(stdErr: string) {
+    const stdErrLines = stdErr.split('\n');
```
Is this safe on Windows?

ChuckJonas replied:
had the same thought... I'm not sure.  Regardless I should use `require('os').EOL` instead

ChuckJonas replied:
@lcampos @vazexqi  What are your thoughts about the line below (for some reason github won't let me start a new comment)?

There's gotta be a better way of handling the update + other junk before the response json.  IMO, it would be easier if the error results were in stdout... 

How have you guys handled this in other commands?

vazexqi commented:
Also, just as an FYI, the "continuous-integration/appveyor/pr " isn't going to pass for external PRs since the autobuild has some encrypted values that are not accessible to external contributors.

As this PR gets further along, we will work out a plan on what to do for CI – we might just bring the branch back instead of a fork.

ChuckJonas commented:
@lcampos 

- moved the error parsing to the utils package
- added unit testing and fixed the unit tests for error parsing
- fixed broken test for SourceDeployCommand
-- I wasn't sure how best to actually test executing the command so I don't have any coverage around the creation of Diagnostics...

ntotten commented:
Related to #397

lcampos commented:
I'm closing this PR since these changes got merged as part of PR #717.
